File: js/src/gdb/tests/test-Root.cpp
#include "gdb-tests.h"

#include "jsapi.h"

#include "gc/Barrier.h"
#include "js/Array.h"  // JS::NewArrayObject
#include "js/GlobalObject.h"
#include "js/ValueArray.h"
#include "vm/JSFunction.h"

using namespace js;

FRAGMENT(Root, null) {
  JS::Rooted<JSObject*> null(cx, nullptr);

  breakpoint();

  use(null);
}

void callee(JS::Handle<JSObject*> obj,
            JS::MutableHandle<JSObject*> mutableObj) {
  // Prevent the linker from unifying this function with others that are
  // equivalent in machine code but not type.
  fprintf(stderr, "Called " __FILE__ ":callee\n");
  breakpoint();
}

FRAGMENT(Root, handle) {
  JS::Rooted<JSObject*> global(cx, JS::CurrentGlobalOrNull(cx));
  callee(global, &global);
  use(global);
}

FRAGMENT(Root, HeapSlot) {
  JS::Rooted<JS::Value> plinth(
      cx, JS::StringValue(JS_NewStringCopyZ(cx, "plinth")));
  JS::Rooted<JSObject*> array(
      cx, JS::NewArrayObject(cx, JS::HandleValueArray(plinth)));

  breakpoint();

  use(plinth);
  use(array);
}

FRAGMENT(Root, barriers) {
  JSObject* obj = JS_NewPlainObject(cx);
  PreBarriered<JSObject*> prebarriered(obj);
  GCPtr<JSObject*> heapptr(obj);
  HeapPtr<JSObject*> relocatable(obj);

  JS::Value val = JS::ObjectValue(*obj);
  PreBarriered<JS::Value> prebarrieredValue(JS::ObjectValue(*obj));
  GCPtr<JS::Value> heapValue(JS::ObjectValue(*obj));
  HeapPtr<JS::Value> relocatableValue(JS::ObjectValue(*obj));

  breakpoint();

  use(prebarriered);
  use(heapptr);
  use(relocatable);
  use(val);
  use(prebarrieredValue);
  use(heapValue);
  use(relocatableValue);
}

File: js/src/gdb/tests/test-Root.py
# Test printing Handles.
# Ignore flake8 errors "undefined name 'assert_pretty'"
# As it caused by the way we instanciate this file
# flake8: noqa: F821

assert_subprinter_registered('SpiderMonkey', 'instantiations-of-JS::Rooted')
assert_subprinter_registered('SpiderMonkey', 'instantiations-of-JS::Handle')
assert_subprinter_registered('SpiderMonkey', 'instantiations-of-JS::MutableHandle')
assert_subprinter_registered('SpiderMonkey', 'instantiations-of-js::BarrieredBase')

run_fragment('Root.handle')

assert_pretty('obj', '(JSObject * const)  [object global] delegate')
assert_pretty('mutableObj', '(JSObject *)  [object global] delegate')

run_fragment('Root.HeapSlot')

# This depends on implementation details of arrays, but since HeapSlot is
# not a public type, I'm not sure how to avoid doing *something* ugly.
assert_pretty('((js::NativeObject *) array.ptr)->elements_[0]', '$JS::Value("plinth")')

run_fragment('Root.barriers')

assert_pretty('prebarriered', '(JSObject *)  [object Object]')
assert_pretty('heapptr', '(JSObject *)  [object Object]')
assert_pretty('relocatable', '(JSObject *)  [object Object]')
assert_pretty('val', '$JS::Value((JSObject *)  [object Object])')
assert_pretty('heapValue', '$JS::Value((JSObject *)  [object Object])')
assert_pretty('prebarrieredValue', '$JS::Value((JSObject *)  [object Object])')
assert_pretty('relocatableValue', '$JS::Value((JSObject *)  [object Object])')
